cirq-aqt/cirq_aqt/aqt_device.py

```diff
@@ -218,13 +223,148 @@ def simulate_samples(self, repetitions: int) -> cirq.Result:
+            # coverage: ignore
+            return [ops.rx(np.pi).on(op.qubits[0]), ops.ry(-1 * np.pi / 2).on(op.qubits[0])]
+        if protocols.has_unitary(op):
+            # coverage: ignore
+            gates = transformers.single_qubit_matrix_to_phased_x_z(protocols.unitary(op))
+            return [g.on(op.qubits[0]) for g in gates]
+        return NotImplemented
+
+    def _decompose_two_qubit_operation(self, op: 'cirq.Operation', _) -> DecomposeResult:
+        if protocols.has_unitary(op):
+            # coverage: ignore
+            return transformers.two_qubit_matrix_to_ion_operations(
+                op.qubits[0], op.qubits[1], protocols.unitary(op)
+            )
+        return NotImplemented
+
+    @property
+    def postprocess_transformers(self) -> List['cirq.TRANSFORMER']:
+        """List of transformers which should be run after decomposing individual operations."""
+        return [transformers.drop_negligible_operations, transformers.drop_empty_moments]
+
+
+@value.value_equality
+class AQTDevice(devices.Device):
     """Ion trap device with qubits having all-to-all connectivity and placed on a line."""
 
+    def __init__(
+        self,
+        measurement_duration: 'cirq.DURATION_LIKE',
+        twoq_gates_duration: 'cirq.DURATION_LIKE',
+        oneq_gates_duration: 'cirq.DURATION_LIKE',
+        qubits: Iterable[devices.LineQubit],
+    ) -> None:
+        """Initializes the description of an ion trap device.
+
+        Args:
+            measurement_duration: The maximum duration of a measurement.
+            twoq_gates_duration: The maximum duration of a two qubit operation.
+            oneq_gates_duration: The maximum duration of a single qubit
+            operation.
+            qubits: Qubits on the device, identified by their x location.
+
+        Raises:
+            TypeError: If not all the qubits supplied are `cirq.LineQubit`s.
+        """
+        self._measurement_duration = value.Duration(measurement_duration)
+        self._twoq_gates_duration = value.Duration(twoq_gates_duration)
+        self._oneq_gates_duration = value.Duration(oneq_gates_duration)
+        if not all(isinstance(qubit, devices.LineQubit) for qubit in qubits):
+            raise TypeError(
+                "All qubits were not of type cirq.LineQubit, instead were "
+                f"{set(type(qubit) for qubit in qubits)}"
+            )
+        self.qubits = frozenset(qubits)
+        self.gateset = AQTTargetGateset()
+
+        graph = nx.Graph()
+        graph.add_edges_from([(a, b) for a in qubits for b in qubits if a != b], directed=False)
+        self._metadata = devices.DeviceMetadata(self.qubits, graph)
+
+    @property
+    def metadata(self) -> devices.DeviceMetadata:
+        return self._metadata
+
+    def decompose_circuit(self, circuit: circuits.Circuit) -> circuits.Circuit:
+        return transformers.optimize_for_target_gateset(circuit, gateset=self.gateset)
+
+    def duration_of(self, operation):
```

Comment:
Didn't we move durations into metadata?
Maybe @MichaelBroughton can advise?

I think main device class is for validation and metadata is for "exploration" (i.e. understanding the topology of the device).

Reply:
These can be moved to metadata yes.